File: ScalaProject/sanity-pipeline/pipelines/LIVY_SCALA/code/src/main/scala/io/prophecy/pipelines/livy_scala/graph/FlattenSchema_1.scala
package io.prophecy.pipelines.livy_scala.graph

import io.prophecy.libs._
import io.prophecy.pipelines.livy_scala.udfs.PipelineInitCode._
import io.prophecy.pipelines.livy_scala.udfs.UDFs._
import io.prophecy.pipelines.livy_scala.config.Context
import org.apache.spark._
import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql.expressions._
import java.time._

object FlattenSchema_1 {

  def apply(context: Context, in: DataFrame): DataFrame = {
    val flattened = in
      .withColumn("arr_str1", explode_outer(col("arr_str1")))
      .withColumn("arr_str2", explode_outer(col("arr_str2")))
      .withColumn("struct_complex-col2",
                  explode_outer(col("struct_complex.col2"))
      )
      .withColumn("struct_complex-col1",
                  explode_outer(col("struct_complex.col1"))
      )
    flattened.select(
      if (flattened.columns.contains("arr_str1")) col("arr_str1")
      else col("arr_str1"),
      if (flattened.columns.contains("arr_str2")) col("arr_str2")
      else col("arr_str2"),
      if (flattened.columns.contains("struct_complex-col2"))
        col("struct_complex-col2")
      else col("struct_complex.col2").as("struct_complex-col2"),
      if (flattened.columns.contains("struct_complex-col1"))
        col("struct_complex-col1")
      else col("struct_complex.col1").as("struct_complex-col1"),
      if (flattened.columns.contains("year")) col("year") else col("year")
    )
  }

}

File: TestPython/pipelines/LIVY_PYTHON/code/livy_python/graph/FlattenSchema_1.py
from pyspark.sql import *
from pyspark.sql.functions import *
from pyspark.sql.types import *
from prophecy.utils import *
from prophecy.libs import typed_lit
from livy_python.config.ConfigStore import *
from livy_python.udfs.UDFs import *

def FlattenSchema_1(spark: SparkSession, in0: DataFrame) -> DataFrame:
    flt_col = in0\
        .withColumn("a2", explode_outer("a2"))\
        .withColumn("a3", explode_outer("a3"))\
        .withColumn("a1", explode_outer("a1"))\
        .withColumn("s1-col2", explode_outer("s1.col2"))\
        .withColumn("s1-col1", explode_outer("s1.col1"))\
        .columns
    selectCols = [col("a1") if "a1" in flt_col else col("a1"),  col("a2") if "a2" in flt_col else col("a2"),                   col("a3") if "a3" in flt_col else col("a3"),                   col("s1-col2") if "s1-col2" in flt_col else col("s1.col2").alias("s1-col2"),                   col("s1-col1") if "s1-col1" in flt_col else col("s1.col1").alias("s1-col1"),                   col("year") if "year" in flt_col else col("year")]

    return in0\
        .withColumn("a2", explode_outer("a2"))\
        .withColumn("a3", explode_outer("a3"))\
        .withColumn("a1", explode_outer("a1"))\
        .withColumn("s1-col2", explode_outer("s1.col2"))\
        .withColumn("s1-col1", explode_outer("s1.col1"))\
        .select(*selectCols)
